server/manager/src/main/java/org/apache/accumulo/manager/upgrade/PreUpgradeValidation.java

```diff
@@ -112,9 +140,163 @@ private static String extractAuthName(ACL acl) {
+        Thread.currentThread().interrupt();
+        invalidLocks.put(lockPath.toString(), "");
+        return;
+      } catch (Exception ex) {
+        log.debug("Failed to establish a connection for to service holding lock. {}", lockPath, ex);
+        invalidLocks.put(lockPath.toString(), "");
+        return;
+      }
+    }
+    log.debug("Failed to establish a connection for to service holding lock. {}. no retires left.",
+        lockPath);
+    invalidLocks.put(lockPath.toString(), "");
+  }
+
+  private List<Pair<HostAndPort,ServiceLock.ServiceLockPath>> gatherLocks(final ZooKeeper zk,
+      final String zkPathRoot) {
+    List<Pair<HostAndPort,ServiceLock.ServiceLockPath>> hosts = new ArrayList<>();
+    try {
+      ZKUtil.visitSubTreeDFS(zk, zkPathRoot, false, (rc, path, ctx, name) -> {
+        if (name.startsWith(ServiceLock.ZLOCK_PREFIX)) {
+          log.trace("found lock at {}", path);
+          try {
+            Stat stat = new Stat();
+            final var zLockPath = ServiceLock.path(path);
+            byte[] lockData = zk.getData(zLockPath.toString(), false, stat);
+            if (lockData == null || !validLockData(lockData, hosts, zLockPath)) {
+              log.debug("Invalid lock data - trying to delete path: {}", zLockPath);
+              zk.delete(zLockPath.toString(), stat.getVersion());
+            }
+          } catch (KeeperException.NoNodeException ex) {
+            // empty - lock no longer exists.
+          } catch (KeeperException ex) {
+            log.trace("could not read lock from ZooKeeper for {}, skipping", path, ex);
+          } catch (InterruptedException ex) {
+            Thread.currentThread().interrupt();
+            throw new IllegalStateException(
+                "interrupted reading lock from ZooKeeper for path: " + path);
+          }
+        }
+      });
+    } catch (InterruptedException ex) {
+      Thread.currentThread().interrupt();
+      throw new IllegalStateException(
+          "interrupted reading lock from ZooKeeper for path: " + zkPathRoot);
+    } catch (KeeperException ex) {
+      log.trace("could not read lock all locks from ZooKeeper for {}, lost list may be incomplete",
+          zkPathRoot, ex);
+    }
+    return hosts;
+  }
+
+  /**
+   * Decode the lock data and if valid add the host, port pair to the hosts map. If the data cannot
+   * be read return false to indicate invalid data.
+   */
+  private boolean validLockData(final byte[] lockData,
+      List<Pair<HostAndPort,ServiceLock.ServiceLockPath>> hosts,
+      ServiceLock.ServiceLockPath zLockPath) {
+    try {
+      Optional<ServiceLockData> sld = ServiceLockData.parse(lockData);
+      if (sld.isPresent()) {
+        HostAndPort hostAndPort = sld.orElseThrow().getAddress(ServiceLockData.ThriftService.TSERV);
+        hosts.add(new Pair<>(hostAndPort, zLockPath));
+      }
+      return true;
+    } catch (Exception ex) {
+      log.trace("Invalid lock data for path: {}", zLockPath, ex);
```

Comment:
Unless this expected for some reason, could log at a higher level.

```suggestion
      log.warn("Invalid lock data for path: {}", zLockPath, ex);
```